**1. Логистическая регрессия**

In [35]:
from keras.datasets import fashion_mnist
from keras.models import Sequential
from keras.layers.core import Dense, Activation
from tensorflow.keras import optimizers
from keras.utils import np_utils
from keras.callbacks import TensorBoard

# Настраиваем параметры сети
NB_EPOCH = 200
BATCH_SIZE = 128
VERBOSE = 1 # Включаем индикатор выполнения
NB_CLASSES = 10 # Количество классов = количеству классов одежды
OPTIMIZER = optimizers.SGD() # Стохастический оптимизатор градиентного спуска
VALIDATION_SPLIT = 0.2 # Определяем какая часть обучающего набора зарезервирована для контроля качества
RESHAPED = 784 # Во входном слое с каждым пикселем изображения ассоциирован 1 нейрон, те. получается 28*28 = 784 нейрона

# Загружаем данные и разбиваем их на обучающий и тестовый набор
(X_train, y_train), (X_test, y_test) = fashion_mnist.load_data()

# Нормализуем данные
X_train = X_train.reshape(60000, RESHAPED)
X_test = X_test.reshape(10000,RESHAPED)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')

X_train /= 255
X_test /= 255

#Преобразовываем векторы классов в бинарные матрицы классов
Y_train = np_utils.to_categorical(y_train, NB_CLASSES)
Y_test = np_utils.to_categorical(y_test, NB_CLASSES)

model = Sequential() # Создаем модел
model.add(Dense(NB_CLASSES, input_shape = (RESHAPED,))) # Добавляем входной полносвязанный слой 
model.add(Activation('softmax')) # Слой активации (Сигмоида)
model.summary()

#Компиляция модели
model.compile(loss = 'categorical_crossentropy', optimizer = OPTIMIZER, metrics = ['accuracy'])



4431872/4422102 [==============================] - 0s 0us/step
Model: "sequential_7"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_5 (Dense)             (None, 10)                7850      
                                                                 
 activation_5 (Activation)   (None, 10)                0         
                                                                 
Total params: 7,850
Trainable params: 7,850
Non-trainable params: 0
_________________________________________________________________


In [36]:
# Подключаем модуль для построения графиков качества модели на валидационной и тренировочной выборке от количества эпох
tensorboard = TensorBoard(log_dir='./logs', histogram_freq=0,
                          write_graph=True, write_images=False)

# Обучаем нашу модель
model.fit(X_train, Y_train, batch_size = BATCH_SIZE, epochs = NB_EPOCH, verbose = VERBOSE, validation_split = VALIDATION_SPLIT, callbacks=[tensorboard])

Epoch 1/200
375/375 [==============================] - 1s 3ms/step - loss: 1.2412 - accuracy: 0.6205 - val_loss: 0.8990 - val_accuracy: 0.7181
Epoch 2/200
375/375 [==============================] - 1s 2ms/step - loss: 0.8273 - accuracy: 0.7363 - val_loss: 0.7591 - val_accuracy: 0.7558
Epoch 3/200
375/375 [==============================] - 1s 2ms/step - loss: 0.7323 - accuracy: 0.7663 - val_loss: 0.6962 - val_accuracy: 0.7762
Epoch 4/200
375/375 [==============================] - 1s 2ms/step - loss: 0.6812 - accuracy: 0.7829 - val_loss: 0.6572 - val_accuracy: 0.7885
Epoch 5/200
375/375 [==============================] - 1s 2ms/step - loss: 0.6471 - accuracy: 0.7932 - val_loss: 0.6303 - val_accuracy: 0.7950
Epoch 6/200
375/375 [==============================] - 1s 2ms/step - loss: 0.6223 - accuracy: 0.7999 - val_loss: 0.6102 - val_accuracy: 0.7999
Epoch 7/200
375/375 [==============================] - 1s 2ms/step - loss: 0.6028 - accuracy: 0.8050 - val_loss: 0.5948 - val_accuracy: 0.8033

In [37]:
# Проверяем модель на тестовых данных
model.evaluate(X_test, Y_test, verbose = VERBOSE)

313/313 [==============================] - 0s 1ms/step - loss: 0.4465 - accuracy: 0.8435


[0.4464704990386963, 0.843500018119812]

In [ ]:
# Строим графики качества модели на валидационной и тренировочной выборке от количества эпох
%load_ext tensorboard
%tensorboard --logdir logs

**Вывод** 

Качество модели по accuracy составило 86.30% на тренировочной выборке и 84.35% на тестовой
